jerry-core/ecma/operations/ecma-jobqueue.c

```diff
@@ -28,26 +34,221 @@
  */
 
 /**
+ * Create a PromiseReactionJob
+ *
+ * @return pointer to the PromiseReactionJob
+ */
+static ecma_job_promise_reaction_t *
+create_promise_reaction_job (ecma_value_t reaction, /**< PromiseReaction */
+                             ecma_value_t argument) /**< argument for the reaction */
+{
+  ecma_job_promise_reaction_t *job_p;
+  job_p = (ecma_job_promise_reaction_t *) jmem_heap_alloc_block (sizeof (ecma_job_promise_reaction_t));
+
+  job_p->reaction = ecma_copy_value (reaction);
+  job_p->argument = ecma_copy_value (argument);
+
+  return job_p;
+} /* create_promise_reaction_job */
+
+/**
+ * Free the heap and the member of the PromiseReactionJob
+ */
+static void
+free_promise_reaction_job (ecma_job_promise_reaction_t *job_p) /**< points to the PromiseReactionJob */
+{
+  JERRY_ASSERT (job_p != NULL);
+
+  ecma_free_value (job_p->reaction);
+  ecma_free_value (job_p->argument);
+
+  jmem_heap_free_block (job_p, sizeof (ecma_job_promise_reaction_t));
+} /* free_promise_reaction_job */
+
+/**
+ * Create a PromiseResolveThenableJob
+ *
+ * @return pointer to the PromiseResolveThenableJob
+ */
+static ecma_job_promise_resolve_thenable_t *
+create_promise_resolve_thenable_job (ecma_value_t promise, /**< promise to be resolved */
+                                     ecma_value_t thenable, /**< thenable object */
+                                     ecma_value_t then) /**< 'then' function */
+{
+  ecma_job_promise_resolve_thenable_t *job_p;
+  job_p = (ecma_job_promise_resolve_thenable_t *) jmem_heap_alloc_block (sizeof (ecma_job_promise_resolve_thenable_t));
+
+  job_p->promise = ecma_copy_value (promise);
+  job_p->thenable = ecma_copy_value (thenable);
+  job_p->then = ecma_copy_value (then);
+
+  return job_p;
+} /* create_promise_resolve_thenable_job */
+
+/**
+ * Free the heap and the member of the PromiseResolveThenableJob
+ */
+static void
+free_promise_resolve_thenable_job (ecma_job_promise_resolve_thenable_t *job_p) /**< points to the
+                                                                                 *  PromiseResolveThenableJob */
+{
+  JERRY_ASSERT (job_p != NULL);
+
+  ecma_free_value (job_p->promise);
+  ecma_free_value (job_p->thenable);
+  ecma_free_value (job_p->then);
+
+  jmem_heap_free_block (job_p, sizeof (ecma_job_promise_resolve_thenable_t));
+} /* free_promise_resolve_thenable_job */
+
+/**
  * The processor for PromiseReactionJob
+ *
+ * See also: ES2015 25.4.2.1
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value
+ */
+static ecma_value_t
+process_promise_reaction_job (void *obj_p) /**< the job to be operated */
+{
+  ecma_job_promise_reaction_t *job_p = (ecma_job_promise_reaction_t *) obj_p;
+  ecma_object_t *reaction_p = ecma_get_object_from_value (job_p->reaction);
+
+  ecma_string_t *str_0 = ecma_new_ecma_string_from_uint32 (0);
+  ecma_string_t *str_1 = ecma_new_ecma_string_from_uint32 (1);
+  ecma_string_t *str_2 = ecma_new_ecma_string_from_uint32 (2);
```

Comment:
In spec, they are "record" instead of object. I use object to represent them so that I can use of the current gc system instead of creating my own.
The relationship between spec definition and my design are list in the first post https://github.com/jerryscript-project/jerryscript/pull/1695#issue-217750183

Only internal code will use those properties. I choose indices to save space and time (create string from c char is less effecient) But I realized maybe I can use **LIT_INTERNAL_MAGIC_STRING_XXX** to represent the name. What do you think about it?